File: internal/tuntap/tun.go
//go:build linux

package tuntap

import (
	"os"
	"syscall"
	"unsafe"

	/*
		#cgo CFLAGS: -Wall -Wpedantic
		#include <stdlib.h>

		int AllocateTun(char* dev);
		int GoErrno();
	*/
	"C"
)

func CAllocateTun(dev string) (*os.File, error) {
	cstr := C.CString(dev)
	defer C.free(unsafe.Pointer(cstr))

	fd := C.AllocateTun(cstr)
	if fd < 0 {
		return nil, syscall.Errno(C.GoErrno())
	}

	return os.NewFile(uintptr(fd), "/dev/net/tun"), nil
}

File: internal/tuntap/tun.c
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/if.h>
#include <linux/if_tun.h>
#include <errno.h>

/**
 * @brief 获取错误码
 *
 * @return int errno
 */
int GoErrno(){
  return errno;
}

/**
 * @brief 创建一个指定的 tun 设备
 *
 * @param dev 设备名
 * @return int 设备fd
 */
int AllocateTun(char *dev) {
  struct ifreq ifr;
  int fd, err;

  if ((fd = open("/dev/net/tun", O_RDWR)) < 0) {
    return -1;
  }

  memset(&ifr, 0, sizeof(ifr));
  ifr.ifr_flags = IFF_TUN;
  if (*dev) {
    strncpy(ifr.ifr_name, dev, IFNAMSIZ);
  }

  if ((err = ioctl(fd, TUNSETIFF, (void *)&ifr)) < 0) {
    close(fd);
    return -1;
  }

  return fd;
}